compiler/noirc_evaluator/src/ssa/validation/mod.rs

```diff
@@ -0,0 +1,472 @@
+//! Validator that checks whether a function is well formed.
+//!
+//! It validates:
+//!
+//! SSA form
+//!
+//! - That the function contains exactly one return block.
+//! - That every checked signed addition or subtraction instruction is
+//!   followed by a corresponding truncate instruction with the expected bit sizes.
+//!
+//! Type checking
+//! - Check that the input values of certain instructions matches that instruction's constraint
+//!   At the moment, only [Instruction::Binary], [Instruction::ArrayGet], and [Instruction::ArraySet]
+//!   are type checked.
+use fxhash::FxHashSet as HashSet;
+
+use crate::ssa::ir::instruction::TerminatorInstruction;
+
+use super::ir::{
+    function::Function,
+    instruction::{Binary, BinaryOp, Instruction, InstructionId, Intrinsic},
+    types::{NumericType, Type},
+    value::{Value, ValueId},
+};
+
+/// Aside the function being validated, the validator maintains internal state
+/// during instruction visitation to track patterns that span multiple instructions.
+struct Validator<'f> {
+    function: &'f Function,
+    // State for truncate-after-signed-sub validation
+    // Stores: Option<(bit_size, result)>
+    signed_binary_op: Option<(u32, ValueId)>,
+}
+
+impl<'f> Validator<'f> {
+    fn new(function: &'f Function) -> Self {
+        Self { function, signed_binary_op: None }
+    }
+
+    /// Validates that any checked signed add/sub is followed by the expected truncate.
+    fn validate_truncate_after_signed_sub(&mut self, instruction: InstructionId) {
```

Comment:
nit: after_signed_add_or_sub?

Reply:
Renamed to `validate_signed_op_overflow_pattern` in https://github.com/noir-lang/noir/pull/8798